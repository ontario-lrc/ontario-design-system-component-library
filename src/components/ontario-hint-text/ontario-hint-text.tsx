import { Component, Prop, Element, h, Watch, State, Method } from '@stencil/core';
import { v4 as uuid } from 'uuid';

import { Hint, HintContentType } from '../../utils/common/common.interface';

import { validatePropExists } from '../../utils/validation/validation-functions';
import { ConsoleMessageClass } from '../../utils/console-message/console-message';

@Component({
	tag: 'ontario-hint-text',
	styleUrl: 'ontario-hint-text.scss',
	shadow: true,
})
export class OntarioHintText implements Hint {
	@Element() host: HTMLElement;

	/**
	 * The content type of the hint.
	 * If no prop is passed, it will default to a string.
	 * If the hint requires multiple lines or HTML, the `hintContentType` prop should be set to `html`.
	 */
	@Prop({ mutable: true }) hintContentType?: HintContentType = 'string';

	/**
	 * Text to display as the hint text statement.
	 *
	 * Setting the hint can be done using the host element textContent or through setting this property.  This property will take precedence.
	 *
	 * @example
	 * <ontario-hint-text hint="Override Hint Text">Hint Text</ontario-button>
	 *
	 * The resulting hint text will display `"Override Hint Text"`.
	 */
	@Prop() hint: string;

	/*
	 * Used to establish a relationship between the hint text content and elements using aria-describedby. This is optional -  if no ID is provided, one will be generated.
	 */
	@Prop({ mutable: true }) elementId?: string;

	@State() hintState: string;

	/**
	 * Watch for changes to the `hintContentType` prop for validation purposes.
	 * If none is provided, or the wrong type is provided, it will default to `string`.
	 */
	@Watch('hintContentType')
	private checkHintContentType() {
		if (this.hintContentType !== 'string' && this.hintContentType !== 'html') {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(' hintContentType ')
				.addRegularText('for')
				.addMonospaceText(' <ontario-hint-text> ')
				.addRegularText('was not one of the permitted types. A default type of `string` will be applied.')
				.printMessage();

			return (this.hintContentType = 'string');
		}

		return this.hintContentType;
	}

	/*
	 * Watch for changes in the `hint` prop for validation purposes.
	 * If no `hint` prop is provided, the `hint` will be set to the host element textContent (if it exists).
	 */
	@Watch('hint')
	private updateHintContent() {
		this.hintState = this.hint ?? this.host.textContent ?? '';
		this.validateHintContent(this.hintState);
	}

	/*
	 * Validate the `hint` and make sure the `hint` has a value.
	 * Log a warning if user doesn't input a value for the `hint` or element content.
	 */
	validateHintContent(newValue: string) {
		// If element content is not provided, check whether prop exists
		if (!this.host.textContent) {
			if (validatePropExists(newValue)) {
				const message = new ConsoleMessageClass();
				message
					.addDesignSystemTag()
					.addMonospaceText(' hint ')
					.addRegularText('for')
					.addMonospaceText(' <ontario-hint-text> ')
					.addRegularText('was not provided')
					.printMessage();
			}
		}
	}

	public getId(): string {
		return this.elementId ?? '';
	}

	/**
	 * This method returns the ontario-hint-text id. It is used to make sure the hint text and `aria-describedby` value of other form components match when the internal hint text props are used.
	 *
	 * @returns Promise<string | undefined>
	 */
	@Method()
	async getHintTextId(): Promise<string | undefined> {
		return this.elementId;
	}

	/**
	 * Set `hint` using internal component logic
	 */
	componentWillLoad() {
		this.updateHintContent();
		this.checkHintContentType();
		this.elementId = this.elementId ?? uuid();
	}

	/**
	 * This helper is used to help load translations for any slots + text content passed in by the user.
	 */
	componentDidLoad() {
		const observer = new MutationObserver((mutations) => {
			mutations.forEach((mutation) => {
				if (mutation.type === 'attributes') {
					this.updateHintContent();
				}
			});
		});

		const options = { attributes: true };
		observer.observe(this.host, options);
	}

	render() {
		return this.hintContentType === 'string' ? (
			<p id={this.getId()} class="ontario-hint">
				{this.hintState}
			</p>
		) : (
			<div id={this.getId()} class="ontario-hint" innerHTML={this.hintState}></div>
		);
	}
}
