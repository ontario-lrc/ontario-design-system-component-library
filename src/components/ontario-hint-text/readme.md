# ontario-hint-text

Use hint text to help users understand how to complete fields in a form. When adding hint text, the first choice should always be to display hint text without using a hint expander.

## Usage guidance

Please refer to the [Ontario Design System](https://designsystem.ontario.ca/components/detail/hint-text.html) for current documentation guidance.

## Configuration

Once the component package has been installed (see Ontario Design System Component Library for installation instructions), the hint text component can be added directly into the project's code, and can be customized by updating the properties outlined [here](#properties). Please see the [examples](#examples) below for how to configure the component.

## Examples

Example of a bare-bones hint text component where the user is passing the `hint` directly as the element text content, which will be displayed as the question/statement.

```html
<ontario-hint-text>This is a hint</ontario-hint-text>
```

Example of a hint text that includes the `hint` property, which will override the `This is a hint` text. The `element-id` property has also been added, which is used to establish a relationship between hint text content and elements using the `aria-describedby` attribute.

```html
<ontario-hint-text hint="This hint will override the hint text provided as element content" element-id="hint-293849">
	This is a hint
	<ontario-hint-text></ontario-hint-text
></ontario-hint-text>
```

<!-- Auto Generated Below -->

## Properties

| Property          | Attribute           | Description                                                                                                                                                                       | Type                              | Default     |
| ----------------- | ------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------- | ----------- |
| `elementId`       | `element-id`        | The unique identifier of the element. This is optional - if no ID is passed, one will be generated.                                                                               | `string \| undefined`             | `undefined` |
| `hint`            | `hint`              | Text to display as the hint text statement. Setting the hint can be done using the host element textContent or through setting this property. This property will take precedence. | `string`                          | `undefined` |
| `hintContentType` | `hint-content-type` | The content type of the hint. If no prop is passed, it will default to a string. If the hint requires multiple lines or HTML, the `hintContentType` prop should be set to `html`. | `"html" \| "string" \| undefined` | `'string'`  |

## Methods

### `getHintTextId() => Promise<string | undefined>`

This method returns the ontario-hint-text id. It is used to make sure the hint text and `aria-describedby` value of other form components match when the internal hint text props are used.

#### Returns

Type: `Promise<string | undefined>`

Promise<string | undefined>

## Dependencies

### Used by

- [ontario-checkboxes](../ontario-checkbox)
- [ontario-dropdown-list](../ontario-dropdown-list)
- [ontario-input](../ontario-input)
- [ontario-radio-buttons](../ontario-radio-buttons)
- [ontario-textarea](../ontario-textarea)

### Graph

```mermaid
graph TD;
  ontario-checkboxes --> ontario-hint-text
  ontario-dropdown-list --> ontario-hint-text
  ontario-input --> ontario-hint-text
  ontario-radio-buttons --> ontario-hint-text
  ontario-textarea --> ontario-hint-text
  style ontario-hint-text fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
