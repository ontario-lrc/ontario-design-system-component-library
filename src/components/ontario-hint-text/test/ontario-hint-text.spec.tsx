import { newSpecPage } from '@stencil/core/testing';
import { OntarioHintText } from '../ontario-hint-text';
import { mutationObserverMock } from '../../../utils/tests/mutation-observer.mock';

global.MutationObserver = mutationObserverMock;

describe('ontario-hint-text', () => {
	describe('snapshot', () => {
		it('should render the expected html', async () => {
			const page = await newSpecPage({
				components: [OntarioHintText],
				html: `<ontario-hint-text element-id="idTest" hint="This is a test hint"></ontario-hint-text>`,
			});

			expect(page.root).toMatchSnapshot();
		});
	});

	it('renders', async () => {
		const page = await newSpecPage({
			components: [OntarioHintText],
			html: `<ontario-hint-text element-id="idTest" hint="This is a test hint"></ontario-hint-text>`,
		});
		expect(page.root).toEqualHtml(`
      <ontario-hint-text element-id="idTest" hint="This is a test hint">
        <mock:shadow-root>
          <p id="idTest" class="ontario-hint">
            This is a test hint
          </p>
        </mock:shadow-root>
      </ontario-hint-text>
    `);
	});
});
