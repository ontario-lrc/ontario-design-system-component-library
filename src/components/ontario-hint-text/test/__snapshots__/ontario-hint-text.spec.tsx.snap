// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ontario-hint-text snapshot should render the expected html 1`] = `
<ontario-hint-text element-id="idTest" hint="This is a test hint">
  <mock:shadow-root>
    <p class="ontario-hint" id="idTest">
      This is a test hint
    </p>
  </mock:shadow-root>
</ontario-hint-text>
`;
